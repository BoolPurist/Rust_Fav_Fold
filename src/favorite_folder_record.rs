use serde::{Deserialize, Serialize};
use std::error::Error;
use std::fmt::Display;
use std::path::Path;

#[derive(Debug, Serialize, Deserialize)]

/// Struct to encapsulate a valid favorite created by the user
/// This favorite is a not empty label with an absolute path
/// That path does need to point to existing location.
pub struct FavoriteFolderPath {
    name: String,
    path: String,
}

impl FavoriteFolderPath {
    /// # Errors
    ///
    /// - If name or path are empty or white spaces.
    /// - If path can not be resolved to a valid utf 8 string.
    /// - If a relative path can not be resolved to an absolute one.
    pub fn new(name: &str, path: &Path) -> Result<Self, InvalidFavoriteFields> {
        let trimmed_name = name.trim().to_string();
        if trimmed_name.is_empty() {
            return Err(InvalidFavoriteFields::EmptyName);
        }

        let abs_path = path
            .canonicalize()
            .map_err(|_| InvalidFavoriteFields::InvalidUtf8PathStr)?;

        let path_str = abs_path
            .to_str()
            .ok_or(InvalidFavoriteFields::InvalidUtf8PathStr)?;

        let trimmed_path = path_str.trim().to_string();

        if trimmed_path.is_empty() {
            return Err(InvalidFavoriteFields::EmptyPath);
        }

        Ok(Self {
            name: trimmed_name,
            path: trimmed_path,
        })
    }

    pub fn set_name(&mut self, new_name: &str) {
        self.name = new_name.to_string()
    }
    pub fn get_name(&self) -> &str {
        &self.name
    }

    pub fn get_path(&self) -> &str {
        &self.path
    }
    pub fn does_exit(&self) -> bool {
        Path::new(&self.path).exists()
    }
}

/// Error to describe the cause of failing to create a struct [`FavoriteFolderPath`]
#[derive(Debug)]
pub enum InvalidFavoriteFields {
    EmptyName,
    EmptyPath,
    InvalidUtf8PathStr,
    UnableAbsPathResolve,
}

impl Display for InvalidFavoriteFields {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            InvalidFavoriteFields::EmptyName => write!(
                f,
                "Name of favorite folder should not be empty or only whitespace"
            ),
            InvalidFavoriteFields::EmptyPath => write!(
                f,
                "Path of favorite folder should not be empty or only whitespace"
            ),
            InvalidFavoriteFields::InvalidUtf8PathStr => {
                write!(f, "Path of favorite folder is not a valid utf 8 text")
            }
            InvalidFavoriteFields::UnableAbsPathResolve => {
                write!(f, "Path could not be resolved to absolute")
            }
        }
    }
}

impl Error for InvalidFavoriteFields {}
