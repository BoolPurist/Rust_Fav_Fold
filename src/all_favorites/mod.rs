use std::path::Path;

use serde::{Deserialize, Serialize};

use crate::favorite_folder_record::FavoriteFolderPath;

#[derive(Debug, PartialEq, Eq, Default, Serialize, Deserialize)]
#[serde(transparent)]
pub struct AllFavorites(Vec<FavoriteFolderPath>);

impl From<AllFavorites> for Vec<FavoriteFolderPath> {
    fn from(value: AllFavorites) -> Self {
        value.0
    }
}

impl AllFavorites {
    pub fn new(folders: impl IntoIterator<Item = FavoriteFolderPath>) -> Self {
        Self(folders.into_iter().collect())
    }

    fn find_by_name(&self, name: &str) -> Option<usize> {
        self.0.iter().position(|fav| fav.get_name() == name)
    }

    pub fn remove_with_name(&mut self, name: &str) -> bool {
        match self.find_by_name(name) {
            Some(index) => {
                self.0.remove(index);
                true
            }
            None => false,
        }
    }

    pub fn clean_all_dangling(&mut self, mut on_is_dangling_check: impl FnMut(&Path) -> bool) {
        self.0
            .retain_mut(|maybe_dangling| !on_is_dangling_check(maybe_dangling.path()));
    }
}

#[cfg(test)]
mod testing {
    use std::path::Path;

    use crate::{favorite_folder_record::FavoriteFolderPath, AllFavorites};

    const INPUT: &str = include_str!("test_input.json");
    fn given() -> AllFavorites {
        let parsed: Vec<FavoriteFolderPath> = serde_json::from_str(INPUT).unwrap();
        AllFavorites::new(parsed)
    }

    #[test]
    fn if_dangling_delete_a_path() {
        let expected = AllFavorites::new([FavoriteFolderPath::new(
            "proc_macro",
            &Path::new("/home/some_user/Code/rust/proc-macro-workshop"),
        )
        .unwrap()]);

        let mut given = given();
        given.clean_all_dangling(|path| {
            let path = path.to_str().unwrap();
            path == "/home/some_user/Documents/Studium" || path == "/home/some_user/Code/rust"
        });
        assert_eq!(expected, given);
    }

    #[test]
    fn find_a_path_by_name() {
        fn assert_case(name: &str, expected: Option<usize>) {
            let given = given();
            let actual = given.find_by_name(name);
            assert_eq!(expected, actual, "Name: {}", name);
        }

        assert_case("a", None);
        assert_case("dev_rust", Some(1));
    }

    #[test]
    fn name_not_found_dont_remove_it() {
        let (actual, has_deleted) = set_up_and_act_remove("not_there");
        const EXPECTED: bool = false;
        assert_eq!(EXPECTED, has_deleted);
        insta::assert_debug_snapshot!(actual);
    }

    #[test]
    fn name_found_then_remove_it() {
        let (actual, has_deleted) = set_up_and_act_remove("dev_rust");
        const EXPECTED: bool = true;
        assert_eq!(EXPECTED, has_deleted);
        insta::assert_debug_snapshot!(actual);
    }

    fn set_up_and_act_remove(name: &str) -> (AllFavorites, bool) {
        let mut given = given();
        let actual = given.remove_with_name(name);
        (given, actual)
    }
}
