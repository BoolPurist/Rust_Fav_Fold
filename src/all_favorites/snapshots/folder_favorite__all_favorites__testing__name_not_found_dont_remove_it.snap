---
source: src/all_favorites/mod.rs
expression: actual
---
AllFavorites(
    [
        FavoriteFolderPath {
            name: "studium",
            path: "/home/some_user/Documents/Studium",
        },
        FavoriteFolderPath {
            name: "dev_rust",
            path: "/home/some_user/Code/rust",
        },
        FavoriteFolderPath {
            name: "proc_macro",
            path: "/home/some_user/Code/rust/proc-macro-workshop",
        },
    ],
)
